src/rust/engine/graph/src/lib.rs

```diff
@@ -188,6 +188,83 @@ impl<N: Node> InnerGraph<N> {
     None
   }
 
+  ///
+  /// Compute the critical path for this graph.
+  ///
+  /// The critical path is the longest path. For a directed acyclic graph, it is equivalent to a
+  /// shortest path algorithm.
+  ///
+  /// Modify the graph we have to fit into the expectations of the Bellman-Ford shortest graph
+  /// algorithm and use that to calculate the critical path.
+  ///
+  fn critical_path<F>(&self, roots: &[N], duration: &F) -> (Duration, Vec<Entry<N>>)
+  where
+    F: Fn(&Entry<N>) -> Duration,
+  {
+    fn duration_into_weight(d: Duration) -> f64 {
+      -(d.as_nanos() as f64)
+    }
+
+    // First, let's map nodes to edges
+    let mut graph = self.pg.filter_map(
+      |_node_idx, node_weight| Some(Some(node_weight)),
+      |edge_idx, _edge_weight| {
+        let source_node = self.pg.raw_edges()[edge_idx.index()].source();
+        self
+          .pg
+          .node_weight(source_node)
+          .map(duration)
+          .map(duration_into_weight)
+      },
+    );
+
+    // Add a single source that's a parent to to all roots
+    let srcs = roots
+      .iter()
+      .filter_map(|n| self.entry_id(&EntryKey::Valid(n.clone())))
+      .cloned()
+      .collect::<Vec<_>>();
+    let src = graph.add_node(None);
+    for node in srcs {
+      graph.add_edge(src, node, 0.);
+    }
+
+    // Add a single destination that's a child from all the previous destinations
```

Comment:
I think that since bellman_ford computes the distance to all other nodes, it isn't necessary to have this "dsts" array: instead, you can run the algorithm, find the minimum/maximum weight in the `weights` array (which would have the index of the node at the end of the longest path), and then use its index to walk the path backward.